[ArrayManager] REF: Implement concat with reindexing

jorisvandenbossche opened the pull request:
xref https://github.com/pandas-dev/pandas/issues/39146 (the concat work item)

This PR implements `concat` for DataFrames using the ArrayManager, and all tests in `pandas/tests/reshape/` already pass. 

Summary of the changes:

* In `internals/concat.py`, I added a `concatenate_block_managers` equivalent for ArrayManagers. This function rather simply reindexes the managers and then concats the arrays (axis=0) or combines them in a single manager (axis=1).
* The main problem with that simple approach is that when reindexing the ArrayManager with the given indexer, we do not yet know the resulting dtype for when a new all-NA column needs to be added (i.e. the common dtype we would find when actually concatting). And we can't make this eg an all-NaN object array, as that would influence the `find_common_type` logic.
* The way I tried to solve this at the moment is by adding a dummy "proxy" object to signal such an all-NA array (the `NullArrayProxy` class), only tracking its length. And when actually concatting, we can replace this with an actual all-NA array after we determined the resulting dtype. This proxy object only lives inside the concatting code (it only gets created when we will call `concat_compat` and inside this function they always get replaced), so the user will never see this. 
* An alternative could be to first determine what the resulting dtype will be in `internals/concat.py` so we can pass this to the `reindex` functionality, and after reindexing do the actual concatenation. But IMO that will be more complex; the concat / find_common_type logic gets involved twice (before and after reindexing), and getting those dtypes for which we want to know the common type is also quite tricky before actually reindexing the manager. 

Basically, this replaces the 600 lines of code of `internals/concat.py` with 25 lines of `concatenate_array_managers` + the `NullArrayProxy` and `_array_from_proxy` code in `dtypes/concat.py` (but I am quite likely still missing some corner cases!)

Note: I did not explicitly test `copy=True/False` behaviour for concat along the columns with this new code.

cc @jreback @jbrockmendel

jorisvandenbossche committed:
[ArrayManager] Implement concat with reindexing

jorisvandenbossche commented on pandas/core/dtypes/concat.py:
```diff
@@ -132,6 +204,75 @@ def is_nonempty(x) -> bool:
     return np.concatenate(to_concat, axis=axis)
 
 
+def concat_arrays(to_concat):
+    """
+    Alternative for concat_compat but specialized for use in the ArrayManager.
+
+    Differences: only deals with 1D arrays (no axis keyword) and does not skip
+    empty arrays to determine the dtype.
+    In addition ensures that all NullArrayProxies get replaced with actual
+    arrays.
```
This `concat_arrays` is basically a copied+adapated version of `concat_compat`. I initially adapted `concat_compat` to cover both cases. 

But that resulted in quite some special cases inside `concat_compat` (checking the proxy objects, while this is not needed for when `concat_compat` is used outside ArrayManager, + not skipping empties for which I added a keyword). 
So in the end I decided to split the changes in a separate function, so both `concat_compat` and `concat_arrays` are more readable / easier to follow, but it certainly gives some duplication.

jorisvandenbossche commented on pandas/core/dtypes/concat.py:
```diff
@@ -21,11 +26,78 @@
+    -------
+    np.ndarray or ExtensionArray
+    """
+    if is_extension_array_dtype(dtype):
+        return dtype.construct_array_type()._from_sequence(
+            [dtype.na_value] * arr.n, dtype=dtype
+        )
+    elif is_datetime64_ns_dtype(dtype):
+        from pandas.core.arrays import DatetimeArray
+
+        return DatetimeArray._from_sequence([NaT] * arr.n, dtype=dtype)
+    elif is_timedelta64_ns_dtype(dtype):
+        from pandas.core.arrays import TimedeltaArray
+
+        return TimedeltaArray._from_sequence([NaT] * arr.n, dtype=dtype)
+    else:
+        if is_integer_dtype(dtype):
+            dtype = "float64"
+            fill_value = np.nan
+        elif is_bool_dtype(dtype):
+            dtype = object
+
+        if fill_value is lib.no_default:
+            fill_value = np.nan
+
+        arr = np.empty(arr.n, dtype=dtype)
+        arr.fill(fill_value)
+        return arr
```
@jbrockmendel do you know if we have similar functionality already existing? 
(There are some similar pieces of code in `internals/concat.py`, but those are quite entangled with the join units / blocks)

jreback replied:
isn't this basically na_value_for_dtype? (then just filling it)

jorisvandenbossche replied:
Hmm, `na_value_for_dtype` is certainly related, but not sure it would directly help here. Because I need to handle the EA / datetime / timedelta case separately anyhow, and then for the last `else: ` block to create a numpy ndarray, it's basically only ever `np.nan` (unless overridden with the keyword)

jbrockmendel replied:
You're right both that there is a lot of very similar code and that it isnt in a form that is easy to reuse.

Parts of this function seem to do the same thing as _dtype_to_na_value.  Other parts look very similar to the `if self.is_na` portion of `JoinUnit.get_reindexed_values`.

FWIW instead of handling the DTA/TDA cases separately, id do them in the numpy block and then call construction.ensure_wrapped_if_datetimelike

jorisvandenbossche replied:
> FWIW instead of handling the DTA/TDA cases separately, id do them in the numpy block and then call construction.ensure_wrapped_if_datetimelike

Ah, that sounds as a good idea, that should simplify this somewhat (and then also makes more sense to re-use `na_value_for_dtype`). Thanks

jorisvandenbossche commented on pandas/core/dtypes/concat.py:
```diff
@@ -21,11 +26,78 @@
+    """
+    if is_extension_array_dtype(dtype):
+        return dtype.construct_array_type()._from_sequence(
+            [dtype.na_value] * arr.n, dtype=dtype
```
We should maybe consider adding a method to the ExtensionArray interface to create an empty/all-NA array of the given dtype, to avoid this rather inefficient "construct from list of NA scalars".

jbrockmendel replied:
IIRC this can break for dtypes that cant hold NA.  the more robust version is 

```
cls = dtype.construct_array_type()
empty = cls._from_sequence([], dtype=dtype)
indexer = -np.ones(arr., dtype=np.intp)
return empty.take(indexer)
```

(in internals concat we use this in one place and when i tried to change it to what you have here something broke, dont remember what off the top of my head)

jorisvandenbossche replied:
Do we have an example dtype that cannot hold NAs?

jbrockmendel replied:
something like Sparse[int]

jorisvandenbossche replied:
Indeed:

```
In [5]: dtype = pd.SparseDtype("int")

In [6]: dtype
Out[6]: Sparse[int64, 0]

In [7]: dtype.construct_array_type()._from_sequence([dtype.na_value] * 5, dtype=dtype)
...
ValueError: Cannot convert non-finite values (NA or inf) to integer
```

Will use your suggested code for now

jorisvandenbossche replied:
I opened https://github.com/pandas-dev/pandas/issues/39776 for the idea of adding a method to the EA interface for creating an empty/all-NA array from a given dtype.

jorisvandenbossche committed:
fix mypy

jorisvandenbossche committed:
pass through allow dups

jreback commented on pandas/core/dtypes/concat.py:
```diff
@@ -21,11 +26,78 @@
     Helper function for `arr.astype(common_dtype)` but handling all special
     cases.
     """
+    if isinstance(arr, NullArrayProxy):
```
can this be at the end of the ifelf

jorisvandenbossche replied:
Not easily because for this case, `arr` has no dtype, and the checks below a dtype is needed

jreback replied:
ok, maybe worthile to annotate code that we potentitally want to change later (e.g. TODO: AM or something)

jreback commented on pandas/core/dtypes/concat.py:
```diff
@@ -21,11 +26,78 @@
+        arr.fill(fill_value)
+        return arr
+
+
```
this should probably be in pandas/core/dtypes/missing.py

jbrockmendel replied:
not dtypes.cast?

jreback replied:
oh yeah thinking: construct_1d_arraylike_from_scalar is just about what you need @jorisvandenbossche ? (or maybe slight mod to get it there)

jreback replied:
maybe add a construct_1d_arraylike_from_na_value

jorisvandenbossche replied:
For now I made it a method of the Proxy class (since it's for now only used for that class, that seemed OK, and makes it a bit nicer in usage)

jreback commented on pandas/core/dtypes/concat.py:
```diff
@@ -132,6 +204,75 @@ def is_nonempty(x) -> bool:
     return np.concatenate(to_concat, axis=axis)
 
 
+def concat_arrays(to_concat):
```
types if you can

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -21,11 +26,78 @@
+            dtype = object
+
+        if fill_value is lib.no_default:
+            fill_value = np.nan
```
see internals.concat._dtype_to_na_value; should this ever be None instead of nan?

jorisvandenbossche replied:
> see internals.concat._dtype_to_na_value; should this ever be None instead of nan?

Here it sometimes needs to be None for datetimelike data. But that doesn't seem to be the use case in internals `_dtype_to_na_value`. Do you have any idea what kind of use case hits that?

jbrockmendel replied:
this code is outdated; is this question still relevant?

jorisvandenbossche replied:
Looking at that code in internals/concat.py, I don't think it is relevant here. The `has_none_blocks` logic basically ensures that `get_reindexed_values` just gets the block itneger values vs uses float as dtype and np.nan as fill value (the None itself is not used as actual fill_value, it's only used to pass as `upcasted_na` to `get_reindexed_values` to signal no upcasting is needed. 

But here, if we are in the NullArrayProxy, we already know that we have missing values to introduce.

jbrockmendel replied:
poor communication on my part.  "is this question still relevant?" was intended to be about your question "Do you have any idea what kind of use case hits that?", not my question in the comment before that

jorisvandenbossche replied:
> poor communication on my part. "is this question still relevant?" was intended to be about your question "Do you have any idea what kind of use case hits that?", not my question in the comment before that

OK ;) Then from my side this is resolved (AFAIU, the None in internals/concat serves a different purpose)

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -132,6 +204,75 @@ def is_nonempty(x) -> bool:
+
+    Differences: only deals with 1D arrays (no axis keyword) and does not skip
+    empty arrays to determine the dtype.
+    In addition ensures that all NullArrayProxies get replaced with actual
```
could NullArrayProxy be used to handle sketchy cases like [this](https://github.com/pandas-dev/pandas/pull/39574/files#diff-0cece63708dac34cbcecb01dabc98083070a9cfbdf7c69d24722e8c363fc39d7R347)?

jorisvandenbossche replied:
That specific test is still skipped here because it's a special case of a DataFrame being appended with a non-aligned Series (in which case the reindexing happens inside the DataFrame.append, instead of inside the concat machinery which knows better how to handle this), but once that is fixed, then yes, this will properly preserve datetime dtype for that test case.

jbrockmendel replied:
i was referring to using this in the non-ArrayManager version, but i think that may no longer be necessary following the most recent updates to #39574

jorisvandenbossche committed:
simplify _array_from_proxy

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
fix creation empty + turn into method

jorisvandenbossche commented:
I pushed some updates.

Given that the review comments are relatively minor points (but useful! to be clear), does that mean that you are OK with the general approach (of using the temporary proxy class)?

This should be ready for another review round.

jorisvandenbossche commented:
I ran the concat/join/merge benchmarks comparing ArrayManager with BlockManager (it's running the benchmarks on this branch with the default of using BlockManager ("before") vs this branch with one addtitional commit changing the default setting to ArrayManager ("after"), so green is slower and red is faster for the ArrayManager):

```diff
$ asv continuous -f 1.01 -b join_merge HEAD~1 HEAD

       before           after         ratio
     [555d7ac1]       [d0c94a51]
     <am-concat~1>       <am-concat>
+        40.4±3ms       93.3±0.3ms     2.31  join_merge.ConcatDataFrames.time_c_ordered(0, True)
+        40.7±3ms       93.2±0.4ms     2.29  join_merge.ConcatDataFrames.time_c_ordered(0, False)
+        44.2±5ms       93.4±0.2ms     2.11  join_merge.ConcatDataFrames.time_f_ordered(0, True)
+        44.6±4ms       93.0±0.2ms     2.09  join_merge.ConcatDataFrames.time_f_ordered(0, False)
+       304±0.8ms        309±0.7ms     1.02  join_merge.MergeCategoricals.time_merge_cat
-         191±2ms        187±0.9ms     0.98  join_merge.MergeAsof.time_by_int('forward', 5)
-       288±0.4ms        282±0.7ms     0.98  join_merge.MergeAsof.time_by_object('nearest', None)
-       289±0.2ms        282±0.1ms     0.98  join_merge.MergeAsof.time_by_object('nearest', 5)
-     18.5±0.06ms      18.0±0.04ms     0.97  join_merge.Merge.time_merge_dataframe_integer_2key(True)
-       206±0.4ms        199±0.4ms     0.97  join_merge.MergeAsof.time_by_object('forward', None)
-       206±0.5ms       199±0.04ms     0.96  join_merge.MergeAsof.time_by_object('forward', 5)
-     6.98±0.09ms      6.60±0.02ms     0.95  join_merge.Merge.time_merge_dataframe_integer_2key(False)
-        90.8±1ms       85.4±0.1ms     0.94  join_merge.MergeAsof.time_by_int('backward', 5)
-         312±3μs          293±1μs     0.94  join_merge.Concat.time_concat_empty_left(0)
-         312±3μs          292±2μs     0.94  join_merge.Concat.time_concat_empty_right(0)
-       102±0.2ms       95.0±0.3ms     0.93  join_merge.MergeAsof.time_by_object('backward', None)
-       102±0.1ms       95.2±0.3ms     0.93  join_merge.MergeAsof.time_by_object('backward', 5)
-      20.2±0.7ms      18.4±0.07ms     0.91  join_merge.Merge.time_merge_2intkey(False)
-         452±3μs        397±0.9μs     0.88  join_merge.Concat.time_concat_mixed_ndims(0)
-      28.2±0.4ms       24.6±0.1ms     0.87  join_merge.Join.time_join_dataframe_index_multi(True)
-         864±3ms          745±5ms     0.86  join_merge.I8Merge.time_i8merge('right')
-         814±3ms         695±10ms     0.85  join_merge.I8Merge.time_i8merge('left')
-      17.9±0.4ms       15.1±0.4ms     0.84  join_merge.Join.time_join_dataframe_index_shuffle_key_bigger_sort(True)
-         193±1μs        161±0.7μs     0.83  join_merge.Concat.time_concat_empty_right(1)
-         194±2μs        161±0.8μs     0.83  join_merge.Concat.time_concat_empty_left(1)
-         796±7μs          662±1μs     0.83  join_merge.Concat.time_concat_mixed_ndims(1)
-        807±10ms          670±5ms     0.83  join_merge.I8Merge.time_i8merge('outer')
-     3.21±0.01ms         2.65±0ms     0.83  join_merge.Merge.time_merge_dataframe_integer_key(True)
-       106±0.3ms       87.5±0.5ms     0.83  join_merge.MergeOrdered.time_merge_ordered
-        810±10ms          667±4ms     0.82  join_merge.I8Merge.time_i8merge('inner')
-      1.54±0.08s          1.26±0s     0.82  join_merge.JoinIndex.time_left_outer_join_index
-      15.1±0.2ms       12.2±0.3ms     0.80  join_merge.Join.time_join_dataframe_index_single_key_small(True)
-        2.74±0ms         2.17±0ms     0.79  join_merge.Merge.time_merge_dataframe_integer_key(False)
-      23.1±0.2ms      17.8±0.08ms     0.77  join_merge.Join.time_join_dataframe_index_multi(False)
-     23.5±0.04ms      18.0±0.03ms     0.77  join_merge.MergeAsof.time_on_uint64('nearest', 5)
-     22.8±0.05ms      17.3±0.07ms     0.76  join_merge.MergeAsof.time_on_uint64('nearest', None)
-     23.2±0.04ms      17.2±0.06ms     0.74  join_merge.MergeAsof.time_on_int('nearest', 5)
-         419±6μs          310±1μs     0.74  join_merge.Append.time_append_homogenous
-     23.0±0.09ms      16.9±0.03ms     0.73  join_merge.MergeAsof.time_on_int32('nearest', 5)
-     22.5±0.04ms      16.5±0.06ms     0.73  join_merge.MergeAsof.time_on_int('nearest', None)
-     22.5±0.07ms      16.4±0.03ms     0.73  join_merge.MergeAsof.time_on_int32('nearest', None)
-     18.8±0.03ms      13.4±0.05ms     0.71  join_merge.MergeAsof.time_on_uint64('forward', 5)
-      15.2±0.1ms      10.8±0.04ms     0.71  join_merge.Join.time_join_dataframe_index_shuffle_key_bigger_sort(False)
-      15.6±0.3ms       11.0±0.1ms     0.71  join_merge.Join.time_join_dataframe_index_single_key_bigger(False)
-     18.5±0.05ms      13.1±0.08ms     0.71  join_merge.MergeAsof.time_on_uint64('forward', None)
-     6.61±0.01ms      4.63±0.02ms     0.70  join_merge.Join.time_join_dataframes_cross(True)
-     17.4±0.04ms      12.0±0.04ms     0.69  join_merge.MergeAsof.time_on_uint64('backward', 5)
-     6.29±0.02ms      4.28±0.01ms     0.68  join_merge.Join.time_join_dataframes_cross(False)
-     17.0±0.07ms      11.5±0.05ms     0.68  join_merge.MergeAsof.time_on_uint64('backward', None)
-     18.7±0.06ms      12.6±0.03ms     0.68  join_merge.MergeAsof.time_on_int('forward', 5)
-      14.5±0.2ms      9.79±0.01ms     0.68  join_merge.Join.time_join_dataframe_index_single_key_small(False)
-      18.4±0.1ms      12.3±0.04ms     0.67  join_merge.MergeAsof.time_on_int32('forward', 5)
-     18.2±0.05ms      12.2±0.04ms     0.67  join_merge.MergeAsof.time_on_int('forward', None)
-     18.1±0.03ms      12.0±0.02ms     0.66  join_merge.MergeAsof.time_on_int32('forward', None)
-     17.6±0.05ms      11.6±0.04ms     0.66  join_merge.MergeAsof.time_on_int32('backward', 5)
-     17.4±0.04ms      11.3±0.03ms     0.65  join_merge.MergeAsof.time_on_int32('backward', None)
-     17.1±0.02ms      11.0±0.06ms     0.64  join_merge.MergeAsof.time_on_int('backward', 5)
-     16.8±0.07ms      10.8±0.04ms     0.64  join_merge.MergeAsof.time_on_int('backward', None)
-       782±0.5ms          491±2ms     0.63  join_merge.MergeCategoricals.time_merge_object
-         977±2μs          596±2μs     0.61  join_merge.Append.time_append_mixed
-     20.9±0.07ms      12.5±0.03ms     0.60  join_merge.Concat.time_concat_small_frames(0)
-     14.5±0.06ms      8.33±0.03ms     0.57  join_merge.Concat.time_concat_small_frames(1)
-         931±3ms          330±3ms     0.35  join_merge.Merge.time_merge_dataframes_cross(True)
-         931±7ms          330±2ms     0.35  join_merge.Merge.time_merge_dataframes_cross(False)
-        58.4±1ms          318±1μs     0.01  join_merge.ConcatDataFrames.time_f_ordered(1, False)
-        55.9±4ms          264±2μs     0.00  join_merge.ConcatDataFrames.time_f_ordered(1, True)
-       91.2±50ms          317±1μs     0.00  join_merge.ConcatDataFrames.time_c_ordered(1, False)
-       90.3±30ms          265±2μs     0.00  join_merge.ConcatDataFrames.time_c_ordered(1, True)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

The results are quite stable on my benchmarking setup (ran the above twice and gave almost same results, also within a run between different parametrizations is quite consistent), so I think they should give a good picture. Some general observations:

- All benchmarks are on par or faster using ArrayManager, *except* for the `concat(  , axis=0)` cases.
- Many of the join/merge/merge_asof cases show some speedup, and I expect that https://github.com/pandas-dev/pandas/pull/39692 should give a further improvement
- Some of the `concat(.., axis=1)` benchmarks give a huge difference (factor "0.00" for after), but this is largely explained by `copy` (the BlockManager based method copies in this case of no reindexing, while the current ArrayManager version doesn't do that (yet)). If I add a copy, it becomes a more modest speedup of 2-3x (instead of >100x).
- The `concat(.., axis=0)` case that shows a 2x slowdown with ArrayManager is the expected case, I think. That specific benchmark is using a DataFrame with a single float dtype (single block), so concatenating 20 2D arrays is always going to be faster as concatenating 200 times 20 1D arrays (the benchmark case has 200 columns, and does `pd.concat([df] * 20, axis=0)`). From profiling the operation, almost all time is spent in the actual numpy concatenation routine (for both cases), so I don't expect there is much room for improvement here.

Generally I am quite positive about the results.

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -5,27 +5,90 @@
+            if is_integer_dtype(dtype):
+                dtype = np.dtype("float64")
+            elif is_bool_dtype(dtype):
+                dtype = np.dtype(object)
```
L71-L74 are basically `maybe_promote(dtype)` (in fact im thinking of implementing a much smaller `ensure_dtype_can_hold_na` that would be almost exclusively this)

jorisvandenbossche replied:
I think in the meantime you added this? (will use it here then)

jorisvandenbossche committed:
remove overriding of fill_value

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche commented:
More comments on this?

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
use ensure_dtype_can_hold_na

jreback commented on pandas/core/dtypes/concat.py:
```diff
@@ -132,6 +183,73 @@ def is_nonempty(x) -> bool:
     return np.concatenate(to_concat, axis=axis)
 
 
+def concat_arrays(to_concat) -> ArrayLike:
```
can you type this List[Any]?

jorisvandenbossche replied:
Yes, that should be fine.

jreback commented on pandas/core/dtypes/concat.py:
```diff
@@ -132,6 +183,73 @@ def is_nonempty(x) -> bool:
 
+def concat_arrays(to_concat) -> ArrayLike:
+    """
+    Alternative for concat_compat but specialized for use in the ArrayManager.
```
if we have different impl for array vs block manager then let's push this code down into those areas (so its localized). ok for a followup.

jreback replied:
the problem is the logic is too scattered and very hard to grok where things are happening.

jreback commented on pandas/core/internals/concat.py:
```diff
@@ -36,6 +36,51 @@
     from pandas.core.arrays.sparse.dtype import SparseDtype
 
 
+def concatenate_array_managers(
```
same comment as above

jreback reviewed (changes requested):
can you do the issue about adding the all-na method first. this PR is adding lots of things that in reality should not be needed if the structure were simplified.

jorisvandenbossche commented:
> can you do the issue about adding the all-na method first. this PR is adding lots of things that in reality should not be needed if the structure were simplified.

I opened https://github.com/pandas-dev/pandas/issues/39776 for this. And to be clear, having that method would replace the following 3 lines of code with a single call to that to-be-defined method:

```
            empty = dtype.construct_array_type()._from_sequence([], dtype=dtype)
            indexer = -np.ones(self.n, dtype=np.intp)
            return empty.take(indexer, allow_fill=True)
```

I don't think that needs to block this PR? 
(but, as said before, I am in favor of adding it to the EA interface. Once there is agreement of how it should look like / what it does do (see the issue), I am happy to implement it)

jorisvandenbossche committed:
add type annotation

jbrockmendel commented:
> More comments on this?

ill take another look today.  i keep waking up to a lot of notifications!

jreback commented:
> > can you do the issue about adding the all-na method first. this PR is adding lots of things that in reality should not be needed if the structure were simplified.
> 
> I opened #39776 for this. And to be clear, having that method would replace the following 3 lines of code with a single call to that to-be-defined method:
> 
> ```
>             empty = dtype.construct_array_type()._from_sequence([], dtype=dtype)
>             indexer = -np.ones(self.n, dtype=np.intp)
>             return empty.take(indexer, allow_fill=True)
> ```
> 
> I don't think that needs to block this PR?
> (but, as said before, I am in favor of adding it to the EA interface. Once there is agreement of how it should look like / what it does do (see the issue), I am happy to implement it)

its fine to do after as well (won't block on that). but normally like to have these precursors first (I know its slightly more of a hassle on the development side), but then we are adding code int he right place and doing smaller, more focused PRs.

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -1,31 +1,82 @@
+    Only stores the length of the array, and not the dtype. The dtype
+    will only be known when actually concatenating (after determining the
+    common dtype, for which this proxy is ignored).
+    Using this object avoids that the internals/concat.py needs to determine
+    the proper dtype and array type.
```
```suggestion

Using this object simplifies the problem (in internals/concat.py) of determining the result dtype of a concatenation.
```

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -132,6 +183,73 @@ def is_nonempty(x) -> bool:
     return np.concatenate(to_concat, axis=axis)
 
 
+def concat_arrays(to_concat: List[Any]) -> ArrayLike:
```
List[Any] -> List[ArrayLike]?

jorisvandenbossche replied:
With the `NullArrayProxy` handling, this is quite annoying to add more specific types

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -132,6 +183,73 @@ def is_nonempty(x) -> bool:
+    single_dtype = len({x.dtype for x in to_concat_no_proxy}) == 1
+    any_ea = any(is_extension_array_dtype(x.dtype) for x in to_concat_no_proxy)
+
+    if any_ea:
```
this looks really similar to the non-AM code.  it cant be shared?

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -322,20 +440,33 @@ def _concat_datetime(to_concat, axis=0):
     a single array, preserving the combined dtypes
     """
     to_concat = [ensure_wrapped_if_datetimelike(x) for x in to_concat]
+    to_concat_no_proxy = [x for x in to_concat if not isinstance(x, NullArrayProxy)]
```
should NullArrayProxy already be excluded by the time we get here?

jbrockmendel replied:
never mind, i see

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -1,31 +1,82 @@
 from pandas.core.construction import array, ensure_wrapped_if_datetimelike
 
 
+class NullArrayProxy:
```
another option if we dont want a whole new thing would be to use `np.empty(shape, dtype="V")`  (which has obj.nbytes = 0)

jbrockmendel commented on pandas/tests/reshape/merge/test_merge.py:
```diff
@@ -670,6 +673,8 @@ def test_join_append_timedeltas(self):
             }
         )
+        if using_array_manager:
+            expected = expected.astype(object)
```
why this change?

jorisvandenbossche replied:
Similar as the comment above: because `df` has object dtype, and for now in this PR I am following the casting rules.

Will add a comment about it.

jbrockmendel commented on pandas/tests/reshape/merge/test_merge.py:
```diff
@@ -277,6 +279,7 @@ def test_merge_copy(self):
         merged["d"] = "peekaboo"
         assert (right["d"] == "bar").all()
 
+    @td.skip_array_manager_invalid_test  # TODO(ArrayManager) join copy behaviour
```
is this invalid or just not implemented?

jbrockmendel commented on pandas/tests/frame/methods/test_join.py:
```diff
@@ -3,15 +3,10 @@
 import pandas._testing as tm
 
-# TODO(ArrayManager) concat with reindexing
-pytestmark = td.skip_array_manager_not_yet_implemented
```
is this fixed by the rest of the PR or just by re-writing the test?

jbrockmendel commented on pandas/tests/frame/methods/test_append.py:
```diff
@@ -183,6 +180,9 @@ def test_append_dtypes(self):
             {"bar": Series([np.nan, Timestamp("20130101")], dtype="M8[ns]")}
         )
+        if using_array_manager:
+            # With ArrayManager, all-NaN float is not ignored
```
can the new JoinUnit.is_valid_na_for be used for these?

jbrockmendel replied:
i think weve landed on the idea being to keep the behaviors matching wherever feasible

jorisvandenbossche replied:
> I think weve landed on the idea being to keep the behaviors matching wherever feasible

In general where it is easy to do so, yes. But for now I kept the new implementation strictly dtype-dependent without any value-dependent behaviour. I don't think we need to mimic every special case or inconsistency of the BM in the new code, but of course in this case it depends on what behaviour we want long term. I opened https://github.com/pandas-dev/pandas/issues/40893 for this.

jbrockmendel commented on pandas/core/internals/array_manager.py:
```diff
@@ -820,10 +835,11 @@ def take(self, indexer, axis: int = 1, verify: bool = True, convert: bool = True
         )
 
-    def _make_na_array(self, fill_value=None):
+    def _make_na_array(self, fill_value=None, use_na_proxy=False):
```
do we need ArrayLike to include NullArrayProxy?

jbrockmendel commented on pandas/core/internals/array_manager.py:
```diff
@@ -739,6 +750,8 @@ def _reindex_indexer(
         allow_dups: bool = False,
         copy: bool = True,
+        do_integrity_check=True,
+        use_na_proxy=False,
```
annotate

jbrockmendel commented on pandas/core/internals/array_manager.py:
```diff
@@ -820,10 +835,11 @@ def take(self, indexer, axis: int = 1, verify: bool = True, convert: bool = True
         if fill_value is None:
             fill_value = np.nan
-
         dtype, fill_value = infer_dtype_from_scalar(fill_value)
```
is the rest of this method ndarray-only?

jbrockmendel commented on pandas/core/internals/array_manager.py:
```diff
@@ -820,10 +835,11 @@ def take(self, indexer, axis: int = 1, verify: bool = True, convert: bool = True
-    def _make_na_array(self, fill_value=None):
+    def _make_na_array(self, fill_value=None, use_na_proxy=False):
+        if use_na_proxy:
+            return NullArrayProxy(self.shape_proper[0])
```
if we have both use_na_proxy and a non-None fill_value does that indicate something went wrong?  i.e should we not silently ignore fill_value here?

jbrockmendel replied:
thanks

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -132,6 +183,73 @@ def is_nonempty(x) -> bool:
+    result = np.concatenate(to_concat)
+
+    # TODO(ArrayManager) this is currently inconsistent between Series and DataFrame
+    # so we should decide whether to keep the below special case or remove it
```
is this behavior consistent between AM and BM?

jorisvandenbossche replied:
Ah, found my notes about this: this is not directly related to Block vs ArrayManager, but something I noticed here that is inconsistent between Series vs DataFrame and empty vs non-empty. 

For Series, with non-empty, we actually coerce to numeric, while if we have empty bool + float, we get object dtype:

```
In [7]: pd.concat([pd.Series([True], dtype=bool), pd.Series([1.0], dtype=float)])
Out[7]: 
0    1.0
0    1.0
dtype: float64

In [8]: pd.concat([pd.Series([], dtype=bool), pd.Series([], dtype=float)])
Out[8]: Series([], dtype: object)
```

While DataFrame uses object dtype in both cases.

(will open a separate issue about it)

But so the reason this came up specifically for ArrayManager, is that we now follow the same rules for Series/DataFrame, and thus also get numeric for non-empty for the DataFrame case.

jorisvandenbossche replied:
Opened https://github.com/pandas-dev/pandas/issues/39817 about this

jorisvandenbossche commented:
@jbrockmendel Thanks for the thorough review!

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
address review

jorisvandenbossche committed:
update comment

jorisvandenbossche committed:
fixup test

jreback commented on pandas/core/dtypes/concat.py:
```diff
@@ -136,6 +187,73 @@ def is_nonempty(x) -> bool:
+        if len(kinds) != 1:
+            if "b" in kinds:
+                result = result.astype(object)
+    return result
```
see my comments above, the non-internals code should ideally not have these hacks here. This makes it *really* hard to follow anything. Can you do something about this.

jorisvandenbossche replied:
Shall I simply cut and paste this function to `internals/concat.py`, and then all is "solved"? (then the null array proxy is not used outside of internals)

(note that my question is slightly ironic, because it doesn't change anything fundamentally, just about semantics what is considered "internal". But I don't really understand what the problem here is).

jreback reviewed (changes requested):
i think we should not merge this and instead fix up w/o the NullArrayProxy thing. This is polluting lots of non-internals code and is likely to take a fair amount of work to remove.

jorisvandenbossche commented:
> This is polluting lots of non-internals code and is likely to take a fair amount of work to remove.

1) This is not polluting "lots of non-internals code". The proxy object is only used *internally* (in the meaning: a user will never see it), and the code to deal with it is currently located in ``core/dtypes/concat.py``. Yes, this is not "located" in `core/internals`, but it still is pandas-internal functionality. If you want that it literally is not used "outside of internals", I am happy to move the `concat_arrays` function that deals with the proxy object to `core/internals` (it's also used *only* there in the internals). 
  Now one point to consider: the `concat_arrays` shares some similarity with `concat_compat`, so for me that is a reason to keep them next to each other. Personally, I think it is perfectly fine to see the proxy object as something "internal" to our concatting code that lives in `core/dtypes/concat.py`.

2) "likely to take a fair amount of work to remove" -> there is no intention to remove this. The null proxy object is meant as the way to deal with reindexed concat in the ArrayManager, and if we eventually decide to keep ArrayManager, we can keep it (or further improve / refactor it if there are better ideas, but no direct reason to remove it)

jorisvandenbossche commented:
Going to punt for now on the `concat` with axis=0 case (for which the proxy object is being used) that generates the discussion. I separated out the minimal pieces of this PR to just get `concat(.., axis=1)` working, which at least already enables merge/join as well, which will be very useful -> https://github.com/pandas-dev/pandas/pull/39841

jreback commented:
maybe i misunderstood your comments

the Null proxy object was meant to be temporary ?

this is not great as a permanent soln and thus -1 on merging

jorisvandenbossche commented:
Can you explain *why* you don't like the proxy object. And to be clear, this will eventually replace the complex concatentation plan / join unit. I don't particularly like *that* code, to be honest.

jorisvandenbossche commented:
(anyway, I would prefer to first focus on  #39841)

jreback commented:
> Can you explain _why_ you don't like the proxy object. And to be clear, this will eventually replace the complex concatentation plan / join unit. I don't particularly like _that_ code, to be honest.

this is a giant hack meant to work around lack of api. let's fix that first.

jorisvandenbossche commented:
Which API is lacking? (I don't understand what your comment is about)

jreback commented:
https://github.com/pandas-dev/pandas/issues/39776 needs to be done first.

jreback commented:
@jorisvandenbossche again my comment on these lengthly PRs. These are going to take a long time to merge. Please do targeted PRs; they are going to make everyone much more productive, esp given the limited review time.

jorisvandenbossche commented:
> #39776 needs to be done first.

That issue has *nothing* to do with the whole discussion going on. Yes, it will reduce the code a little bit in this PR (to be exact, with **2** lines, see https://github.com/pandas-dev/pandas/pull/39612#issuecomment-778226069 above showing the 3 lines of code that would be replaced by 1 line). But that issue doesn't change the *use* of the NullArrayProxy object at all. And as I understand you have a problem with that in general, not with the 3 lines of code in its implementation to convert itself to an actual array.

Also, if you see that issue so much as a blocker, can you take a look at it and give feedback?

jreback commented:
the NullProxy is complete technical debt. Please find a way w/o doing this.

jbrockmendel commented:
1) @jorisvandenbossche is right that #39776 should not be a blocker for this.  It will allow 2-3 lines of cleanup as a follow-up.

2) If NullArrayProxy can be used to as an _alternative_ to JoinUnit/ConcatenationPlan that'd be great.  To really be a home run, it should be made useful for both AM and BM.

3) Big picture, I continue to think that (but mostly try to keep to myself) that ArrayManager will produce little net simplification because it will necessitate workarounds outside of core.internals (frame.values discussed in earlier threads/calls, #39692, haven't decided on #39772/#39820).  If point 2) is addressed, then this would unequivocally not be subject to this criticism.

> Shall I simply cut and paste this function to internals/concat.py, and then all is "solved"? (then the null array proxy is not used outside of internals) [...] (note that my question is slightly ironic, because it doesn't change anything fundamentally, just about semantics what is considered "internal". But I don't really understand what the problem here is).

You are right this is not A Big Deal.  Nevertheless I'd say "yes", I'd rather see only-ever-used-in-internals code live in internals.concat, or even a AM-specific modules (unless point 2 makes it re-usable for BM)

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
update/remove skips

jorisvandenbossche committed:
move logic into internals

jorisvandenbossche commented:
Time to get back to this. I updated this PR with latest master, removed a bunch of skips that were no longer necessary in the meantime, and moved the full implementation to the internals as asked above (so pandas/core/dtypes/concat.py isn't touched anymore). 

@jreback I know you said that I needed to "find a way without doing this [NullProxy]", but you also replied "I agree with @jbrockmendel points here", where Brock said:

> If NullArrayProxy can be used to as an *alternative* to JoinUnit/ConcatenationPlan that'd be great. 

Indeed, that's the whole point of the NullArrayProxy, to be an alternative (better, more understandable) implementation to JoinUnit/ConcatenationPlan. 

Of course, Brock also said:

>  To really be a home run, it should be made useful for both AM and BM.

In theory, that would be nice, yes. But I would argue that this is not the highest priority at this point, and that it's certainly out of scope for this PR (this PR is already big enough, if we reallly want this, it can be done in a follow-up PR).

A large part of the "concatenation plan" is inherently tied to the BlockManager (dealing with blocks, blklocs, blkno, etc), and cannot be shared with the new code for ArrayManager (it's simply not relevant for ArrayManager). Potentially, a part of the implementation of JoinUnit (the part check which dtype to use etc) could be shared. But note that some helper functions, eg `ensure_dtype_can_hold_na`, are already shared by both.

jorisvandenbossche committed:
fix typing

jbrockmendel commented on pandas/core/internals/array_manager.py:
```diff
@@ -1338,10 +1339,12 @@ def to_array(self, dtype: DtypeObj) -> ArrayLike:
         np.ndarray or ExtensionArray
         """
         if is_extension_array_dtype(dtype):
```
isinstance(dtype, ExtensionDtype)

jbrockmendel commented on .github/workflows/ci.yml:
```diff
@@ -161,7 +161,7 @@ jobs:
         pytest pandas/tests/groupby/
         pytest pandas/tests/resample/
-        pytest pandas/tests/reshape/merge
+        pytest pandas/tests/reshape/
```
nice!  let's use a glob pattern here and get a bunch of these down to a single line

jbrockmendel replied:
youve got pandas/tests/generic/test_generic.py and pandas/tests/generic/ below

jorisvandenbossche replied:
This is replaced by `run_tests.sh` now in anther PR, so marking as resolved.

jbrockmendel commented on pandas/core/dtypes/concat.py:
```diff
@@ -30,7 +30,7 @@
 
 
-def _cast_to_common_type(arr: ArrayLike, dtype: DtypeObj) -> ArrayLike:
+def cast_to_common_type(arr: ArrayLike, dtype: DtypeObj) -> ArrayLike:
```
yah this probably belongs in dtypes.cast (independent of this PR)

jbrockmendel commented on pandas/core/internals/concat.py:
```diff
@@ -92,6 +101,93 @@ def _concatenate_array_managers(
+
+    Parameters
+    ----------
+    to_concat : list of arrays
```
the annotation is List[Any].  can that be made more specific?

jorisvandenbossche replied:
In principle the more specific annotation would be `List[Union[np.ndarray, ExtensionArray, NullArrayProxy]]`. But, iwhen adding that, that gives a bunch of errors down the line because mypy cannot figure out that we are eg converting all input to EAs before passing to `_concat_same_type`. 
So unless if you have specific suggestions how to solve this, I'd rather leave it less specified (for example `concat_compat` is also not typed).

jbrockmendel replied:
i think the practice in this case is to not annotate rather than use Any, cc @simonjayhawkins

jorisvandenbossche replied:
But `List[Any]` at least indicates it needs to receive a list, which can be better than no typing at all?

jbrockmendel replied:
im hoping @simonjayhawkins will weigh in here, but my understanding of the rule of thumb is that `List[Any]` denotes "this cannot be further narrowed down" whereas `List` denotes "its a list and the annotation is incomplete; contributions welcome"

jorisvandenbossche replied:
Ah, I thought you meant to have no annoation at all (like `concat_compat`). I can certainly make it `List` instead of `List[Any]`

simonjayhawkins replied:
I _think_ that's the general consensus at the moment until we add `disallow_any_generics` #30539 which will force us to add type parameters.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
update type check

jorisvandenbossche committed:
simply casting to_concat + fix skips

jorisvandenbossche committed:
further simplify concat_arrays

jorisvandenbossche commented:
@jreback @jbrockmendel any other comments? (apart from the typing ones)

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
remove redundant cast

jorisvandenbossche commented:
If there are no more comments, I am planning to merge this.

jreback commented on pandas/core/internals/concat.py:
```diff
@@ -75,14 +82,16 @@ def _concatenate_array_managers(
         # TODO(ArrayManager) doesn't yet preserve the correct dtype
         arrays = [
-            concat_compat([mgrs[i].arrays[j] for i in range(len(mgrs))])
+            concat_arrays([mgrs[i].arrays[j] for i in range(len(mgrs))])
```
can you now remove concat_compat?

jreback replied:
and reading your comment below, why is this not in array manger if its only used there?

jorisvandenbossche replied:
> can you now remove concat_compat?

`concat_compat` is used in several other places as well

> why is this not in array manger if its only used there?

Because this is the code for concatting managers, which for BlockManager also resides in internals/concat.py?

jbrockmendel commented:
> If there are no more comments, I am planning to merge this.

Again, please do not.  I've given up asking on things that are Sufficiently Benign, but this is not in that category.

jorisvandenbossche commented:
> Again, please do not. I've given up asking on things that are Sufficiently Benign, but this is not in that category.

Brock, I didn't merge this without asking. I know this is a larger PR that I can't just merge, but so that's exactly the reason I asked it above, so you could reply to it before I would actually merge (which you did now ;)).  
(maybe my formulation was not that good, I just wanted to trigger a "review again or indicate it's good")

But to the point: can you then indicate that you are still planning to review in more detail or still have concerns? Because now your last comments were all minor things, which gives me the impression that you no longer have those.

jbrockmendel commented:
Looking at JoinUnit, it does have `self.block`, but it wouldn't be too hard to have it take block.values instead, at which point it would be BlockManager/ArrayManager-agnostic.  `_get_mgr_concatenation_plan` accesses `mgr.blklocs` and `mgr.blknos` twice each, but i think the ArrayManager analogues of those are `np.ones(len(self.arrays))` and `np.arange(len(self.arrays))`

So I think we can adapt existing code (and as a result, behavior) to be shared rather than add a bunch of new code

jorisvandenbossche commented:
> `_get_mgr_concatenation_plan` accesses `mgr.blklocs` and `mgr.blknos` twice each

It indeed accesses those twice, but then the rest of the function further processes those .. AFAIK this whole function is fully BlockManager-specific (also the return value includes BlockPlacement, so is BlockManager-specific)

> Looking at JoinUnit, it does have `self.block`, but it wouldn't be too hard to have it take block.values instead, at which point it would be BlockManager/ArrayManager-agnostic.

It's probably indeed possible to make JoinUnit directly hold the values instead of the Block. But what's your concrete idea here? That it would be used instead of the NullProxy?

Personally, I think the implementation I made right now for ArrayManager is much cleaner and simpler to follow. And my impression is that @jbrockmendel agrees with that (*"If NullArrayProxy can be used to as an alternative to JoinUnit/ConcatenationPlan that'd be great"*). 
If we don't want to have the *temporary* duplication between ArrayManager and BlockManager, then there are two options:

* I ditch my new impementation and rewrite it using bits and pieces of the current BlockManager implementation. Personally, I would not like to do that. The new version is simpler (also because the required logic for ArrayManager is inherently less complex as it doens't need to deal with possibly unaligned blocks that get concatenated), and adding logic for the ArrayManager to the current implementation will only make the already complex piece of code more complex (and at the point when we move to ArrayManager as default, we would want this simpler version anyway).

* We refactor the full of concatenation_plan / JoinUnit based on my NullProxy idea. In principle, I think this would be a good idea, as I think it could mean an improvement of the current implementation. 
  However, this is a quite big refactor, and I would argue that this is 1) not a priority short-term for getting the ArrayManager feature-complete (and be able to do a full evaluation / comparison with the BlockManager) and 2) given that (assuming we decide to go with ArrayManager) the BlockManager is not intended to be actively developed for a long time (eg only 1-2 years), a big refactor might not be worth it, as any refactor always has the risk to introduce regressions (for long lived code that is typically worth it, though). Even if we want to do this, I think this can be a *follow-up* to this PR.

jbrockmendel commented on pandas/core/internals/concat.py:
```diff
@@ -93,6 +102,80 @@ def _concatenate_array_managers(
+        else:
+            target_dtype = to_concat_no_proxy[0].dtype
+    elif not single_dtype:
+        if any(kind in ["m", "M"] for kind in kinds):
```
why cant we use find_common_type here?

jorisvandenbossche replied:
Good question. So, can we think of any combination of dtypes where `np.find_common_type(..)` (which would get used by our find_common_type in this case) would not result in object dtype if the input contains datetime or timedelta?

Most obvious case I could think of is with integers, and numpy also gives object dtype there:

```
In [20]: np.find_common_type([np.dtype("datetime64[ns]"), np.dtype("int64")], [])
Out[20]: dtype('O')
```

The reason I handled it separately here is because in `dtypes/concat.py` we also handle datetime-likes separately, and in `_concat_datetime` we also "hardcode" to use object dtype in case of multiple input dtypes. 
But if `find_common_dtype` always gives object dtype for mixed datetimelike input, this can indeed be simplified here. Will try that and see what the tests say.

jorisvandenbossche replied:
@jbrockmendel so I was indeed able to simplify this, also the `if any_ea` check was not needed, since `find_common_type`  handles all cases

jbrockmendel commented on pandas/core/internals/concat.py:
```diff
@@ -93,6 +102,80 @@ def _concatenate_array_managers(
+
+    result = np.concatenate(to_concat)
+
+    # TODO decide on exact behaviour (we shouldn't do this only for empty result)
```
is this decision a blocker for anything?

jorisvandenbossche replied:
No, this is existing behaviour, unrelated to ArrayManager/BlockManager. I just noticed the inconsistency while implementing this, and opened the issue for it (and for now mimicked the current behaviour in the code below)

jbrockmendel replied:
thanks

jreback commented:
I am ok with temporary duplication of things. This is a pretty common pattern, e.g. first to make things work, then moving towards removing the duplication in steps.

jbrockmendel commented:
Agreed.  I'd like to see the [find_common_type](https://github.com/pandas-dev/pandas/pull/39612#discussion_r602641131) thing resolved first, but wont hold this up based on that

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
simplify usage of find_common_type

jorisvandenbossche commented:
Updated this PR, and resolved the find_common_type comment. The py37_32bit failing test seems a flaky hypothesis one

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
update annotation

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche commented:
Any more comments?

jbrockmendel commented:
> Any more comments?

https://github.com/pandas-dev/pandas/pull/39612/files#r602641371, but doesnt need to be addressed here.  LGTM

jreback commented:
lgtm

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche committed:
fixup typing

jbrockmendel commented:
LGTM, might do another rebase to be on the safe side

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into am-concat

jorisvandenbossche commented:
> https://github.com/pandas-dev/pandas/pull/39612/files#r602641371, but doesnt need to be addressed here. LGTM

Sorry, I missed that one earlier (GitHub's UI is not fantastic when replying to an older comment, as it doesn't show up in the recent comments in the timeline ..). I opened https://github.com/pandas-dev/pandas/issues/40893 to track this inconsistency.
